### PR TITLE
FIX #62 (and #49 ^^)

### LINKED ISSUE #62: Make metadata work in windows again

### PR DESCRIPTION
**Pull request by rderimay**
Added a test for Windows environment to call the exif Windows version.
Improved the path which *should* work with spaces.
Changed the way the MetaDataDialog is created allowing to pass
different width and line counts
Moved splitForColumns to Utils and removed needed code

### CODE REVIEW

File: focuspoints.lrdevplugin/ExifUtils.lua [hunk 1]
```diff
@@ -38,7 +46,7 @@ function ExifUtils.readMetaData(targetPhoto)
   local cmd, metaDataFile = ExifUtils.getExifCmd(targetPhoto)
   LrTasks.execute(cmd)
   local fileInfo = LrFileUtils.readFile(metaDataFile)
-  LrFileUtils.delete(metaDataFile)
+  --LrFileUtils.delete(metaDataFile)
```
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/ExifUtils.lua]: the file needs to get deleted. uncomment. bug

File: focuspoints.lrdevplugin/Utils.lua [hunk 2]
```diff
@@ -31,7 +31,7 @@ local myLogger = LrLogger( 'libraryLogger' )
 myLogger:enable( "logfile" )
 
 isDebug = false
-isLog = false
+isLog = true
```
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/Utils.lua]: false on commits

File: focuspoints.lrdevplugin/MetaDataDialog.lua [hunk 3]
```diff
@@ -26,34 +26,39 @@ local LrPathUtils = import 'LrPathUtils'
...
-  }
-  
-  local row = viewFactory:row {
-    myText, myText2, 
-    margin_left = 5, 
-  }
-  
-  local scrollView = viewFactory:scrolled_view {
-    row, 
+  local f = LrView.osFactory()
```
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/MetaDataDialog.lua]: viewFactory

File: focuspoints.lrdevplugin/Utils.lua [hunk 4]
```diff
@@ -68,7 +68,35 @@ function split(str, delim)
...
 --[[
- Splits a string into 2 parts: key and value. 
+-- Split a key/value table and returns 2 strings being the ordered list of keys and the ordered list of value
+-- as well as the max key length, the max value length and the number of lines in both strings
+-- tab - the table
+--]]
+function splitForColumns(tab)
+  local keys = ""
+  local values = ""
+  local keys_max_length = 0
```
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/Utils.lua]: naming needs to be consistent.

File: focuspoints.lrdevplugin/Utils.lua [hunk 5]
```diff
@@ -126,6 +154,32 @@ function arrayKeyOf(table, val)
 end
 
 --[[
+-- Return a table iteraor which iterates through the table keys in alphabetical order
```
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/Utils.lua]: spelling "iterator"

**Comment by musselwhizzle** [comment]
I'm happy enough with the current metadata viewer than I'll pull this into branch first to test before merging into master.

File: focuspoints.lrdevplugin/Utils.lua (hunk 2 above)
Comment (rderimay) [inline focuspoints.lrdevplugin/Utils.lua]: ok, maybe at some point we could make it in the lr setting page...

File: focuspoints.lrdevplugin/Utils.lua (hunk 4 above)
Comment (rderimay) [inline focuspoints.lrdevplugin/Utils.lua]: I canot name it table as I use table.insert after, if this is what you mean.
We can/should rename table to tab in the other methods thus...

File: focuspoints.lrdevplugin/Utils.lua (hunk 5 above)
Comment (rderimay) [inline focuspoints.lrdevplugin/Utils.lua]: yep

**Comment by rderimay** [comment]
the current metadata viewer is not working at all on windows. this one is working.
No changes on mac (just the size of the left column is "adapted" the lightroom way: too big.

File: focuspoints.lrdevplugin/Utils.lua (hunk 4 above)
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/Utils.lua]: ```keysMaxLength``` no underscores.

File: focuspoints.lrdevplugin/ShowMetadata.lua [hunk 6]
```diff
@@ -31,78 +31,44 @@ require "Utils"
...
+        -- Starting to show progess dialog
+        dialogScope:setIndeterminate()
 
-          dialogScope:done()
-          MetaDataDialog.labels.title = column1
-          MetaDataDialog.data.title = column2
-          --MetaDataDialog.labels.title = "parts: "  .. parts[1].key
-        end)
+        local metaData = ExifUtils.readMetaDataAsTable(targetPhoto)
+        local keywords, values, keywords_max_length, values_max_length, line_count = splitForColumns(metaData)
```
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/ShowMetadata.lua]: rename variables to follow naming convention.

File: focuspoints.lrdevplugin/Utils.lua (hunk 2 above)
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/Utils.lua]: that's the plan at some point.

File: focuspoints.lrdevplugin/ShowMetadata.lua [hunk 7]
```diff
@@ -31,78 +31,44 @@ require "Utils"
...
-          local column1, column2 = splitForColumns(metaData)
+        -- Starting to show progess dialog
+        dialogScope:setIndeterminate()
 
-          dialogScope:done()
-          MetaDataDialog.labels.title = column1
-          MetaDataDialog.data.title = column2
-          --MetaDataDialog.labels.title = "parts: "  .. parts[1].key
-        end)
+        local metaData = ExifUtils.readMetaDataAsTable(targetPhoto)
```
Comment (musselwhizzle) [inline focuspoints.lrdevplugin/ShowMetadata.lua]: as mentioned, this will remove any duplicates which is not what is desired.

**Comment by musselwhizzle** [comment]
too much refactoring for personal preference. You're welcome to leave the code as close to as-is or just let me take the feature when I have time.

**Comment by musselwhizzle** [comment]
PR https://github.com/musselwhizzle/Focus-Points/issues/49 resolved this

**Comment by musselwhizzle** [comment]
In this PR and other, you refactor our code for personal preference. I wont accept code where this happens (such as here https://github.com/musselwhizzle/Focus-Points/blob/4089fb473f8a0b5ebfa5982f6aaf618d6127dbea/focuspoints.lrdevplugin/FocusPointDialog.lua) 

If you wish to contribute (and I'm glad you are) please don't refactor my code for your personal preference.

**Comment by rderimay** [comment]
Should I have wanted to refactor to my "personal convenience" I would have done it an other way.

**Comment by rderimay** [comment]
@musselwhizzle, please be so kind and do not make statements on me doing stuff for "personal preference", please. 
I am literally doing *nothing* for my personal preference here.
I am contributing to an -awesome- open sourced project under Apache license that I like.
Thus, I don't see my code as mine but as ours. I am just trying to make the code better when I think I can bring something. As far as I can evaluate it, I think my contributions were not so devastating to your project right? Should they have been, let me know and I will let you alone.
You totally have the right to -and should use the right of- accepting code or not. In fact, most maintainers do not enough filter out bad stuff in my opinion, leading to an odd patchwork of code. 
I have no problem of you asking me to remove stuff for a good reason and discussing it constructively. In fact, this is very interesting to have someone else have an external look on what you were writing. I see this as a way of improving my coding.
But if you are really taking open source contributors with the respect you should, you should not write what you wrote in the last commit.

**Comment by musselwhizzle** [comment]
You have called code in this project "crap" being offense to other contributors and to my self. You have repeated refactored my code for your personal preference. Example, you changed this code from 

my code
```
  local viewFactory = LrView.osFactory()
  local myPhoto = viewFactory:catalog_photo {
    width = photoW, 
    height = photoH,
    photo = targetPhoto,
  }
  local myText = viewFactory:static_text {
    title = "",
  }
      
  local column = viewFactory:column {
    myPhoto, myText,
  }
  
  local myView = viewFactory:view {
    column, overlayView, 
    place = 'overlapping', 
  }
```

to your code
```
local f = LrView.osFactory()

  local view = f:view {
    f:column {
      f:catalog_photo {
        width = photoWidth,
        height = photoHeight,
        photo = targetPhoto,
      },
      f:static_text {
        title = "" -- "CL " .. developSettings["CropLeft"] .. ", CT " .. developSettings["CropTop"] .. ", Angle " .. developSettings["CropAngle"],
      },
    },

    overlayView,

    place = "overlapping"
  }
```
You can not justify that change (or others) other than for personal preference. 

The project is opened source but that does not mean I must to take contributions from others, including yourself. I said it above and I'll say it again: 

> If you wish to contribute (and I'm glad you are) please don't refactor my code for your personal preference.

If that doesn't sit well with you, you are welcome to no longer contribute.

**Comment by rderimay** [comment]
This is where you dont understand me. I am not saying this was *needed* and I accept you not wanting to merge it. This is **NO** problem.
But you must accept that I did not made it by personal preference. I made it because I thought it was good for the project. In this particular example I proposed the change because I think this is better readable.
About the use of "crap": I never used "crap". Never. 
And I did it, this was not my intent (and I would like you to find where I did it) to offensive. 
If I would have found this project to be uninteresting or crap, of course I would have left without notice.

**Comment by musselwhizzle** [comment]
https://github.com/musselwhizzle/Focus-Points/pull/10

6 days ago in your comments. You called my code crap and then continually refactor my code for style. Think about how well that is going to sit with me.....

**Comment by musselwhizzle** [comment]
I'm happy you're contributing; I'm am. Let's just try not to step on each others toes.

**Comment by rderimay** [comment]
That's the point, this really not my intend.
you mean "crap" in this comment. "Alternatively we could change the Fujidelegate to handle it but I am the opinion this is crap." ???

> Yes I own a few in fact (x-t1, x-t2, x30).
Yes no problem, I am happy to help. In fact I spent most of the time trying to find out how to add my changes to the git :-)
I can also help on other themes should you need capacity. My knowledge on .lua is very limited but it seems to be pretty straight forward.
I understand that your AFpoint is not the center of the focus box but the top left. This explains why I had some deviations.
What do you think ? Isn't it better to change your code to make it the middle of the box for all cameras ? This would require to change the few configuration files that already exist but I feel it to be more future proof to do it now.
Alternatively we could change the Fujidelegate to handle it but I am the opinion this is crap.
For example, the Fuji cameras (I assume many hybrids) can change the size of the focus points (very small, small, normal, big) I already searched for an exif field that could lead top the size but did not find it. I am pretty confident this exists in the exif and I will find it later.
For this reason, it seems more logical to center everything ?

**Comment by musselwhizzle** [comment]
It's ok, really. Perhaps I misunderstood. Perhaps you can be careful not to call my work or your work or nobody's that. I'm glad your here. Let's move on and forget about all of this. You're a good dev and as I've said, I'm happy you're here.

**Comment by rderimay** [comment]
Deal !

**Comment by musselwhizzle** [comment]
do you have any issues with me deleting the past several comments here? I dont want any future press or other contributors getting a wrong or bad impression.

**Pull request closed without merging**